### PR TITLE
[liblzma] Fix for wasm32-emscripten

### LINKED ISSUE #17750: [liblzma] build failure with wasm32-emscripten
**Host Environment**
- OS: Windows
- Compiler: emcc 2.0.20

**To Reproduce**
`./vcpkg install liblzma:wasm32-emscripten`

**Failure logs** 
```
[1/2] cmd /c "cd ../../wasm32-emscripten-dbg && "C:/Users/Bruno/vcpkg/downloads/tools/cmake-3.20.1-windows/cmake-3.20.1-windows-i386/bin/cmake.exe" "C:/Users/Bruno/vcpkg/buildtrees/liblzma/src/v5.2.5-77236e0e89.clean" "-DCMAKE_DEBUG_POSTFIX=d" "-DCMAKE_MAKE_PROGRAM=C:/Users/Bruno/vcpkg/downloads/tools/ninja/1.10.1-windows/ninja.exe" "-DCMAKE_SYSTEM_NAME=Emscripten" "-DBUILD_SHARED_LIBS=OFF" "-DVCPKG_CHAINLOAD_TOOLCHAIN_FILE=C:/Users/Bruno/emsdk/upstream/emscripten/cmake/Modules/Platform/Emscripten.cmake" "-DVCPKG_TARGET_TRIPLET=wasm32-emscripten" "-DVCPKG_SET_CHARSET_FLAG=ON" "-DVCPKG_PLATFORM_TOOLSET=external" "-DCMAKE_EXPORT_NO_PACKAGE_REGISTRY=ON" "-DCMAKE_FIND_PACKAGE_NO_PACKAGE_REGISTRY=ON" "-DCMAKE_FIND_PACKAGE_NO_SYSTEM_PACKAGE_REGISTRY=ON" "-DCMAKE_INSTALL_SYSTEM_RUNTIME_LIBS_SKIP=TRUE" "-DCMAKE_VERBOSE_MAKEFILE=ON" "-DVCPKG_APPLOCAL_DEPS=OFF" "-DCMAKE_TOOLCHAIN_FILE=C:/Users/Bruno/vcpkg/scripts/buildsystems/vcpkg.cmake" "-DCMAKE_ERROR_ON_ABSOLUTE_INSTALL_DESTINATION=ON" "-DVCPKG_CXX_FLAGS=" "-DVCPKG_CXX_FLAGS_RELEASE=" "-DVCPKG_CXX_FLAGS_DEBUG=" "-DVCPKG_C_FLAGS=" "-DVCPKG_C_FLAGS_RELEASE=" "-DVCPKG_C_FLAGS_DEBUG=" "-DVCPKG_CRT_LINKAGE=dynamic" "-DVCPKG_LINKER_FLAGS=" "-DVCPKG_LINKER_FLAGS_RELEASE=" "-DVCPKG_LINKER_FLAGS_DEBUG=" "-DVCPKG_TARGET_ARCHITECTURE=wasm32" "-DCMAKE_INSTALL_LIBDIR:STRING=lib" "-DCMAKE_INSTALL_BINDIR:STRING=bin" "-D_VCPKG_ROOT_DIR=C:/Users/Bruno/vcpkg" "-D_VCPKG_INSTALLED_DIR=C:/Users/Bruno/vcpkg/installed" "-DVCPKG_MANIFEST_INSTALL=OFF" "-G" "Ninja" "-DCMAKE_BUILD_TYPE=Debug" "-DCMAKE_INSTALL_PREFIX=C:/Users/Bruno/vcpkg/packages/liblzma_wasm32-emscripten/debug""
FAILED: ../../wasm32-emscripten-dbg/CMakeCache.txt
cmd /c "cd ../../wasm32-emscripten-dbg && "C:/Users/Bruno/vcpkg/downloads/tools/cmake-3.20.1-windows/cmake-3.20.1-windows-i386/bin/cmake.exe" "C:/Users/Bruno/vcpkg/buildtrees/liblzma/src/v5.2.5-77236e0e89.clean" "-DCMAKE_DEBUG_POSTFIX=d" "-DCMAKE_MAKE_PROGRAM=C:/Users/Bruno/vcpkg/downloads/tools/ninja/1.10.1-windows/ninja.exe" "-DCMAKE_SYSTEM_NAME=Emscripten" "-DBUILD_SHARED_LIBS=OFF" "-DVCPKG_CHAINLOAD_TOOLCHAIN_FILE=C:/Users/Bruno/emsdk/upstream/emscripten/cmake/Modules/Platform/Emscripten.cmake" "-DVCPKG_TARGET_TRIPLET=wasm32-emscripten" "-DVCPKG_SET_CHARSET_FLAG=ON" "-DVCPKG_PLATFORM_TOOLSET=external" "-DCMAKE_EXPORT_NO_PACKAGE_REGISTRY=ON" "-DCMAKE_FIND_PACKAGE_NO_PACKAGE_REGISTRY=ON" "-DCMAKE_FIND_PACKAGE_NO_SYSTEM_PACKAGE_REGISTRY=ON" "-DCMAKE_INSTALL_SYSTEM_RUNTIME_LIBS_SKIP=TRUE" "-DCMAKE_VERBOSE_MAKEFILE=ON" "-DVCPKG_APPLOCAL_DEPS=OFF" "-DCMAKE_TOOLCHAIN_FILE=C:/Users/Bruno/vcpkg/scripts/buildsystems/vcpkg.cmake" "-DCMAKE_ERROR_ON_ABSOLUTE_INSTALL_DESTINATION=ON" "-DVCPKG_CXX_FLAGS=" "-DVCPKG_CXX_FLAGS_RELEASE=" "-DVCPKG_CXX_FLAGS_DEBUG=" "-DVCPKG_C_FLAGS=" "-DVCPKG_C_FLAGS_RELEASE=" "-DVCPKG_C_FLAGS_DEBUG=" "-DVCPKG_CRT_LINKAGE=dynamic" "-DVCPKG_LINKER_FLAGS=" "-DVCPKG_LINKER_FLAGS_RELEASE=" "-DVCPKG_LINKER_FLAGS_DEBUG=" "-DVCPKG_TARGET_ARCHITECTURE=wasm32" "-DCMAKE_INSTALL_LIBDIR:STRING=lib" "-DCMAKE_INSTALL_BINDIR:STRING=bin" "-D_VCPKG_ROOT_DIR=C:/Users/Bruno/vcpkg" "-D_VCPKG_INSTALLED_DIR=C:/Users/Bruno/vcpkg/installed" "-DVCPKG_MANIFEST_INSTALL=OFF" "-G" "Ninja" "-DCMAKE_BUILD_TYPE=Debug" "-DCMAKE_INSTALL_PREFIX=C:/Users/Bruno/vcpkg/packages/liblzma_wasm32-emscripten/debug""
-- Check if the system is big endian
-- Searching 16 bit integer
-- Looking for sys/types.h
-- Looking for sys/types.h - found
-- Looking for stdint.h
-- Looking for stdint.h - found
-- Looking for stddef.h
-- Looking for stddef.h - found
-- Check size of unsigned short
-- Check size of unsigned short - done
-- Searching 16 bit integer - Using unsigned short
-- Check if the system is big endian - TEST_BIG_ENDIAN found no result!
CMake Error at C:/Users/Bruno/vcpkg/downloads/tools/cmake-3.20.1-windows/cmake-3.20.1-windows-i386/share/cmake-3.20/Modules/TestBigEndian.cmake:130 (message):
  TEST_BIG_ENDIAN found no result!
Call Stack (most recent call first):
  C:/Users/Bruno/vcpkg/downloads/tools/cmake-3.20.1-windows/cmake-3.20.1-windows-i386/share/cmake-3.20/Modules/TestBigEndian.cmake:37 (__TEST_BIG_ENDIAN_LEGACY_IMPL)
  cmake/tuklib_integer.cmake:21 (test_big_endian)
  CMakeLists.txt:126 (tuklib_integer)


-- Performing Test HAVE___BUILTIN_BSWAPXX
-- Performing Test HAVE___BUILTIN_BSWAPXX - Success
-- Performing Test HAVE___BUILTIN_ASSUME_ALIGNED
-- Performing Test HAVE___BUILTIN_ASSUME_ALIGNED - Success
-- Looking for clock_gettime
-- Looking for clock_gettime - found
-- Looking for CLOCK_MONOTONIC
-- Looking for CLOCK_MONOTONIC - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE
-- Looking for pthread_condattr_setclock
-- Looking for pthread_condattr_setclock - found
-- Checking how to detect the number of available CPU cores
-- Performing Test TUKLIB_CPUCORES_SCHED_GETAFFINITY
-- Performing Test TUKLIB_CPUCORES_SCHED_GETAFFINITY - Failed
-- Performing Test TUKLIB_CPUCORES_CPUSET
-- Performing Test TUKLIB_CPUCORES_CPUSET - Failed
-- Looking for sys/param.h
-- Looking for sys/param.h - found
-- Performing Test TUKLIB_CPUCORES_SYSCTL
-- Performing Test TUKLIB_CPUCORES_SYSCTL - Failed
-- Performing Test TUKLIB_CPUCORES_SYSCONF
-- Performing Test TUKLIB_CPUCORES_SYSCONF - Success
-- Checking how to detect the amount of physical memory
-- Performing Test TUKLIB_PHYSMEM_SPECIAL
-- Performing Test TUKLIB_PHYSMEM_SPECIAL - Failed
-- Performing Test TUKLIB_PHYSMEM_AIX
-- Performing Test TUKLIB_PHYSMEM_AIX - Failed
-- Performing Test TUKLIB_PHYSMEM_SYSCONF
-- Performing Test TUKLIB_PHYSMEM_SYSCONF - Success
-- Looking for immintrin.h
-- Looking for immintrin.h - found
-- Performing Test HAVE__MM_MOVEMASK_EPI8
-- Performing Test HAVE__MM_MOVEMASK_EPI8 - Failed
-- Looking for getopt_long
-- Looking for getopt_long - found
-- Looking for program_invocation_name
-- Looking for program_invocation_name - found
-- Looking for mbrtowc
-- Looking for mbrtowc - found
-- Looking for wcwidth
-- Looking for wcwidth - found
-- Looking for optreset
-- Looking for optreset - found
-- Looking for posix_fadvise
-- Looking for posix_fadvise - found
-- Performing Test HAVE_STRUCT_STAT_ST_ATIM_TV_NSEC
-- Performing Test HAVE_STRUCT_STAT_ST_ATIM_TV_NSEC - Success
-- Looking for futimens
-- Looking for futimens - found
-- Configuring incomplete, errors occurred!
```

**Additional context**
It seems to have something to do with cmake not being able to determine the endianness.
Apparently cmake has a 
`CMAKE_<LANG>_BYTE_ORDER` variable that can be used to specify it, but I am not familiar with vcpkg and I am not sure how to forward it to the cmake invocation.
Wasm is always little endian, so maybe this can be hardcoded somewhere?

### PR DESCRIPTION
**Pull request by Semphriss**
This PR allows building liblzma with Emscripten by disabling a buggy endianness test.

**Note:** This patch uses a workaround by manually setting the endianness to little-endian; it has been proposed that vcpkg should use an endianness test designed specifically designed for WebAssembly.

- #### What does your PR fix?
  Fixes #17750

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?
  wasm32-emscripten, I couldn't find liblzma in the baseline and I'm not sure if I should add it or not.

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
  Yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?
  Yes

**If you are still working on the PR, open it as a Draft: https://github.blog/2019-02-14-introducing-draft-pull-requests/**

### CODE REVIEW

File: ports/liblzma/emscripten_little_endian.patch [hunk 1]
```diff
@@ -0,0 +1,32 @@
+diff --git a/cmake/tuklib_integer.cmake b/cmake/tuklib_integer.cmake
+index d7e2e28..a674877 100644
+--- a/cmake/tuklib_integer.cmake
++++ b/cmake/tuklib_integer.cmake
+@@ -8,7 +8,9 @@
+ #
+ 
+ include("${CMAKE_CURRENT_LIST_DIR}/tuklib_common.cmake")
+-include(TestBigEndian)
++if(NOT EMSCRIPTEN)
```
Comment (JackBoosY) [inline ports/liblzma/emscripten_little_endian.patch:10]: Where is `EMSCRIPTEN` declared?

**Review by JackBoosY — CHANGES_REQUESTED** [review]

**Comment by dg0yt** [comment]
The bug report indicates an alternative without patching: pass "-DCMAKE_C_BYTE_ORDER=LITTLE_ENDIAN" as cmake parameter, for emscripten LE triplets. vcpkg uses CMake 3.20 so this must be available.

File: ports/liblzma/emscripten_little_endian.patch (hunk 1 above)
Comment (Semphriss) [reply ports/liblzma/emscripten_little_endian.patch:10]: It is automatically declared and set to true by the Emscripten compiler toolchain.

**Comment by Semphriss** [comment]
TIL about that option.

As I wrote in the initial comment, this patch is probably not the most optimal solution; it uses a few checks for an Emscripten compiler (using the Emscripten-defined CMake option, as @JackBoosY asked), but it has been suggested in #17750 that it should be on Emscripten's end to provide an appropriate override for testing endianness. Since I'm not sure if vcpkg uses a custom fork of Emscripten or if it strictly follows the original, I assumed that this would require action from the Emscripten team, and that in the meantime, forcing the option would be the way to go, since all WASM is little-endian anyways.

Whenever I'll have some free time, I'll try to change this PR to use the CMake option instead of hardcoding it; I'll try to search how to do that.

**Comment by JackBoosY** [comment]
I think we should not modify this issue in the port part, because this is a common issue. We need to let the upstream fix part of the cmake tool chain code (_Emscripten.cmake_), or fix it in our tool chain (_triplets/community/wasm32-emscripten.cmake_).

**Comment by dg0yt** [comment]
The issue comments also mention that "emsdk has its own version is TestBigEndian.cmake." From that POV, the vcpkg toolchain perhaps just needs to make to adjust the cmake module path?

**Comment by Semphriss** [comment]
Letting Emscripten fix it is a possibility, but my concern was about the time it would take for it to be implemented (the issue was opened back in May). I wanted to provide a temporary solution so at least vcpkg can work properly while waiting for the issue to be fixed upstream. But I agree that a patch might be a heavy and unrealiable solution.

Should I do the fix directly in the toolchain and open a different PR?

**Comment by dg0yt** [comment]
The wasm32-emscripten triplet cares of chainloading `$ENV{EMSDK}/upstream/emscripten/cmake/Modules/Platform/Emscripten.cmake`. There is:

~~~
list(APPEND CMAKE_MODULE_PATH "${EMSCRIPTEN_ROOT_PATH}/cmake/Modules")
~~~
https://github.com/emscripten-core/emscripten/blob/b2473ea8f53cae1df599108f5d0ae5b6c1a2c599/cmake/Modules/Platform/Emscripten.cmake#L77

From this, it should already use emscripten's `TestBigEndian.cmake` instead of CMake's module.
Why doesn't this just work?

**Comment by JackBoosY** [comment]
This PR has been inactive for a long time, please reopen it if there is any progress.

**Pull request closed without merging**